src/charm.py

```diff
@@ -117,14 +127,87 @@ def _on_config_changed(self, _):
                 self._setup_cron_job(self._stored.config)
         self._update_status()
 
+    def _check_packages(self):
+        # Find all packages that are defined in the "Packages" file of the
+        # mirror_path and snapshot-*. These packages are still in referenced
+        # and should not be removed.
+        indexed_packages = set()
+        snapshot_paths = self._list_snapshots()
+        mirror_path = "{}/mirror".format(self._stored.config["base-path"])
+        for path in snapshot_paths + [mirror_path]:
+            for base, indices in locate_package_indices(path):
+                indexed_packages |= set(find_packages_by_indices(indices, base=base))
+
+        # Find all .deb packages that exist in the mirror_path. It might
+        # contains some packages that are not longer being referenced. We can
+        # clean up those packages using the clean-up-packages action.
+        existing_packages = set(
+            [p.absolute() for p in Path(mirror_path).rglob("**/*.deb")]
+        )
+
+        # Main calculation
+        packages_to_be_removed = existing_packages - indexed_packages
+        freed_up_space = 0
+        for p in packages_to_be_removed:
+            freed_up_space += p.stat().st_size
+        return packages_to_be_removed, convert_bytes(freed_up_space)
+
+    def _on_check_packages_action(self, event):
+        packages_to_be_removed, freed_up_space = self._check_packages()
+        event.set_results(
+            {
+                "message": (
+                    "The following packages can be removed since they have no "
+                    "reference in the current index and indices in all the snapshots."
+                ),
+                "count": len(packages_to_be_removed),
+                "packages": json.dumps(
+                    [str(p) for p in packages_to_be_removed], indent=4
+                ),
+                "total-size": freed_up_space,
+            }
+        )
+
+    def _on_clean_up_packages_action(self, event):
+        if not event.params["confirm"]:
+            logger.info(
+                "clean up action not performed because the user did not confirm "
+                "the action."
+            )
+            event.set_results(
+                {"message": "Aborted! Please confirm your action with 'confirm=true'."}
+            )
+            return
+
+        start = time.time()
+        logger.info("Cleaning up unreferenced packages")
+        packages_to_be_removed, freed_up_space = self._check_packages()
+        for package in packages_to_be_removed:
+            package.unlink()
```

Comment:
I suspect that the action will look "successful", but actually failed silently because `set_result` will not be excuted. I think you are right, we should add some error handling and logs here (there's only error handling for subprocess error).